<a href="https://colab.research.google.com/github/brook-miller/mbai-417-data/blob/main/data-models-databases/homework/homework1-answers.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Homework 1: Promoting Evening Products to Drive Incremental Revenue at Instacart

  

Your team has been brought in as a consultant working with the merchandising team at Instacart.  Early evening hours have delivery / driver availability but the number of orders falls off in the evening hours.  
&nbsp;   
We need to determine which products are being sold in the evening 6PM+ (18-23 in order_hour_of_day) hours so that the team can create new promotions.  To avoid developing a list of niche products we'd like to restrict our analysis to products that only get at least 5000 total orders.
&nbsp;  
The merchandising team has the option of sending out emails or featuring products to shoppers on site, which do you recommend will be more effective in meeting the goal.  Feel free to call out additional data that would be helpful in evaluating the opportunity.  
&nbsp;  
All the data you need is stored in the redshift database.  Please update the code in each cell to build to the final answer.  
&nbsp;  
In the last markdown cell, please discuss your key findings and what next steps and data should be gathered in order to determine how the team should proceed to evaluate the opportunity.  
&nbsp;  
Please limit your analysis to 250 words.  

# Setting up the environment

In [ ]:
#@title installs for sqlalchemy and sqlmagic
!pip install sqlalchemy-redshift --quiet
!pip install redshift_connector --quiet
!pip install ipython-sql --quiet

     |████████████████████████████████| 94 kB 1.9 MB/s 
     |████████████████████████████████| 131 kB 11.4 MB/s 
     |████████████████████████████████| 97 kB 5.8 MB/s 
     |████████████████████████████████| 8.5 MB 36.2 MB/s 
     |████████████████████████████████| 6.4 MB 44.2 MB/s 
     |████████████████████████████████| 503 kB 39.6 MB/s 
     |████████████████████████████████| 79 kB 6.1 MB/s 
     |████████████████████████████████| 138 kB 41.6 MB/s 
     |████████████████████████████████| 127 kB 44.1 MB/s 
     |████████████████████████████████| 104 kB 53.8 MB/s 
ERROR: pip's dependency resolver does not currently take into account all the packages that are installed. This behaviour is the source of the following dependency conflicts.
datascience 0.10.6 requires folium==0.2.1, but you have folium 0.8.3 which is incompatible.


In [ ]:
#@title standard imports - we'll use in most EDA
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import plotly.express as px

from datetime import datetime, timedelta
from dateutil.parser import parse
from google.colab import data_table
data_table.enable_dataframe_formatter()

In [ ]:
#@title setting up sql connection and sql magic, unique to this lab

import getpass
import sqlalchemy as sa
from sqlalchemy.engine.url import URL
from sqlalchemy import orm as sa_orm

connect_to_db = URL.create(
drivername='redshift+redshift_connector',     
host='mbai417-redshift.cuvtmrb8eogw.us-west-2.redshift.amazonaws.com',     
port=5439,  
database='dev',  
username='ro_user',         #username should not be hard coded either
password=getpass.getpass()  #please don't put passwords into code
)

engine = sa.create_engine(connect_to_db)
%reload_ext sql
%sql $connect_to_db

··········


'Connected: ro_user@dev'

# Determining evening products

## Show the number of orders by hour of day

In [ ]:
%%sql
select order_hour_of_day as hour, count(*) from orders
group by hour
order by hour

 * redshift+redshift_connector://ro_user:***@mbai417-redshift.cuvtmrb8eogw.us-west-2.redshift.amazonaws.com:5439/dev
Done.


[(0, 22758),
 (1, 12398),
 (2, 7539),
 (3, 5474),
 (4, 5527),
 (5, 9569),
 (6, 30529),
 (7, 91868),
 (8, 178201),
 (9, 257812),
 (10, 288418),
 (11, 284728),
 (12, 272841),
 (13, 277999),
 (14, 283042),
 (15, 283639),
 (16, 272553),
 (17, 228795),
 (18, 182912),
 (19, 140569),
 (20, 104292),
 (21, 78109),
 (22, 61468),
 (23, 40043)]

## Determine the number of products ordered by hour of day


In [ ]:
%%sql
select o.order_hour_of_day as hour, count(*) from order_products op
inner join orders o on op.order_id = o.order_id
group by hour
order by hour


 * redshift+redshift_connector://ro_user:***@mbai417-redshift.cuvtmrb8eogw.us-west-2.redshift.amazonaws.com:5439/dev
Done.


[(0, 218948),
 (1, 115786),
 (2, 69434),
 (3, 51321),
 (4, 53283),
 (5, 88062),
 (6, 290795),
 (7, 891937),
 (8, 1719973),
 (9, 2456713),
 (10, 2764426),
 (11, 2738582),
 (12, 2620847),
 (13, 2663292),
 (14, 2691548),
 (15, 2664533),
 (16, 2537458),
 (17, 2089465),
 (18, 1637923),
 (19, 1259401),
 (20, 977038),
 (21, 796370),
 (22, 634734),
 (23, 402620)]

## Calculate the basket size (# products / # orders) by hour of day

In [ ]:
%%sql 

with ordercount as (
    select order_hour_of_day as hour, count(*) as ordercount from orders
    group by hour
), productcount as (
    select o.order_hour_of_day as hour, count(*) as productcount from order_products op
    inner join orders o on op.order_id = o.order_id
    group by hour
)
select o.hour, ordercount, productcount, productcount::float / ordercount as basketsize
from ordercount o
inner join productcount p on p.hour = o.hour
order by hour


 * redshift+redshift_connector://ro_user:***@mbai417-redshift.cuvtmrb8eogw.us-west-2.redshift.amazonaws.com:5439/dev
Done.


[(0, 22758, 218948, 9.6207048071008),
 (1, 12398, 115786, 9.339086949507985),
 (2, 7539, 69434, 9.20997479771853),
 (3, 5474, 51321, 9.375411033978809),
 (4, 5527, 53283, 9.640492129545866),
 (5, 9569, 88062, 9.202842512279235),
 (6, 30529, 290795, 9.525205542271284),
 (7, 91868, 891937, 9.708897548656768),
 (8, 178201, 1719973, 9.651870640456563),
 (9, 257812, 2456713, 9.52908708671435),
 (10, 288418, 2764426, 9.584790130990438),
 (11, 284728, 2738582, 9.618239161585795),
 (12, 272841, 2620847, 9.605766728607504),
 (13, 277999, 2663292, 9.58022151158817),
 (14, 283042, 2691548, 9.509359035054867),
 (15, 283639, 2664533, 9.394099542023488),
 (16, 272553, 2537458, 9.309961732213551),
 (17, 228795, 2089465, 9.132476671255928),
 (18, 182912, 1637923, 8.954704994751575),
 (19, 140569, 1259401, 8.959308240081384),
 (20, 104292, 977038, 9.368292870018793),
 (21, 78109, 796370, 10.195624063808268),
 (22, 61468, 634734, 10.326251057460793),
 (23, 40043, 402620, 10.054691206952526)]

## Find the most commonly ordered products

In [ ]:
%%sql

select product_name as product, min(p.product_id) as product_id, count(*) as count from order_products op
    inner join orders o on o.order_id = op.order_id
    inner join products p on p.product_id = op.product_id
    group by product
    order by count(*) desc
    limit 20

 * redshift+redshift_connector://ro_user:***@mbai417-redshift.cuvtmrb8eogw.us-west-2.redshift.amazonaws.com:5439/dev
Done.


[('Banana', 24852, 472565),
 ('Bag of Organic Bananas', 13176, 379450),
 ('Organic Strawberries', 21137, 264683),
 ('Organic Baby Spinach', 21903, 241921),
 ('Organic Hass Avocado', 47209, 213584),
 ('Organic Avocado', 47766, 176815),
 ('Large Lemon', 47626, 152657),
 ('Strawberries', 16797, 142951),
 ('Limes', 26209, 140627),
 ('Organic Whole Milk', 27845, 137905),
 ('Organic Raspberries', 27966, 137057),
 ('Organic Yellow Onion', 22935, 113426),
 ('Organic Garlic', 24964, 109778),
 ('Organic Zucchini', 45007, 104823),
 ('Organic Blueberries', 39275, 100060),
 ('Cucumber Kirby', 49683, 97315),
 ('Organic Fuji Apple', 28204, 89632),
 ('Organic Lemon', 5876, 87746),
 ('Apple Honeycrisp Organic', 8277, 85020),
 ('Organic Grape Tomatoes', 40706, 84255)]

## Find the most commonly ordered products in the evening hours (18-23)

In [ ]:
%%sql
    select product_name as product, min(p.product_id) as product_id, count(*) as count from order_products op
    inner join orders o on o.order_id = op.order_id
    inner join products p on p.product_id = op.product_id
    where o.order_hour_of_day between 18 and 24
    group by product
    order by count(*) desc
    limit 20

 * redshift+redshift_connector://ro_user:***@mbai417-redshift.cuvtmrb8eogw.us-west-2.redshift.amazonaws.com:5439/dev
Done.


[('Banana', 24852, 114571),
 ('Bag of Organic Bananas', 13176, 91171),
 ('Organic Strawberries', 21137, 65604),
 ('Organic Baby Spinach', 21903, 59221),
 ('Organic Hass Avocado', 47209, 52178),
 ('Organic Avocado', 47766, 43132),
 ('Large Lemon', 47626, 35072),
 ('Strawberries', 16797, 33860),
 ('Organic Whole Milk', 27845, 33763),
 ('Organic Raspberries', 27966, 33301),
 ('Limes', 26209, 32564),
 ('Organic Yellow Onion', 22935, 25336),
 ('Organic Blueberries', 39275, 25293),
 ('Organic Garlic', 24964, 25231),
 ('Organic Zucchini', 45007, 25210),
 ('Cucumber Kirby', 49683, 24337),
 ('Organic Lemon', 5876, 21075),
 ('Organic Grape Tomatoes', 40706, 20953),
 ('Organic Fuji Apple', 28204, 20668),
 ('Seedless Red Grapes', 4920, 19871)]

## Calculate products that have the highest percentage of sales in the evening hours

In [ ]:
%%sql
with eveningproducts as (   
    select product_name as product, min(p.product_id) as product_id, count(*) as count from order_products op
    inner join orders o on o.order_id = op.order_id
    inner join products p on p.product_id = op.product_id
    where o.order_hour_of_day between 18 and 24
    group by product
   ), 
   allproducts as (
    select product_name as product, min(p.product_id) as product_id, count(*) as count from order_products op
    inner join orders o on o.order_id = op.order_id
    inner join products p on p.product_id = op.product_id
    group by product
   )
select e.product, e.count as eveningcount, a.count as totalcount, 
(e.count::float) / (totalcount) eveningpercent from eveningproducts e
inner join allproducts a on a.product_id = e.product_id
where totalcount > 5000
order by eveningpercent desc
limit 20

 * redshift+redshift_connector://ro_user:***@mbai417-redshift.cuvtmrb8eogw.us-west-2.redshift.amazonaws.com:5439/dev
Done.


[('Americone Dream® Ice Cream', 1490, 6155, 0.2420796100731113),
 ('Milk and Cookies Ice Cream', 1208, 5003, 0.2414551269238457),
 ('Birthday Cake Light Ice Cream', 1230, 5129, 0.23981282901150322),
 ('Half Baked® Ice Cream', 1409, 5928, 0.23768556005398112),
 ('Strawberry Ice Cream', 1402, 6129, 0.22874857236090718),
 ('Ataulfo Mango', 1270, 5569, 0.2280481235410307),
 ('Mediterranean Mint Gelato', 1650, 7319, 0.22544063396638886),
 ('Chocolate Chip Cookie Dough Ice Cream', 1833, 8137, 0.22526729752980215),
 ('Mini Seedless Cucumbers', 1691, 7559, 0.22370683952903822),
 ('Sea Salt Caramel Gelato', 1476, 6671, 0.22125618348073753),
 ('Mint Chip', 1211, 5502, 0.22010178117048346),
 ('Low Fat Plain Yogurt', 1303, 5934, 0.21958206943040107),
 ('Caramel Cookie Crunch Gelato', 1203, 5517, 0.21805328983143013),
 ('Yobaby Organic Plain Yogurt', 1813, 8337, 0.2174643157010915),
 ('Organic Cherries', 1205, 5629, 0.21406999467045656),
 ('Medium Scarlet Raspberries', 1078, 5037, 0.21401627953146715),
 ('Whole Milk Plain Yogurt', 3258, 15267, 0.21340145411672234),
 ('Organic Baby Spinach Salad', 3741, 17672, 0.2116908103214124),
 ('YoBaby Peach Pear Yogurt', 1820, 8641, 0.2106237703969448),
 ('Organic Peeled & Cooked Beets', 1157, 5512, 0.2099056603773585)]

# Final analysis
In the markdown cell below summarize your findings.  Which products have the highest proportion of sales in evening hours?  Should the merchandising team focus on email or on-site featuring to drive incremental orders.  What additional data would you collect to evaluate the opportunity? Please limit your response to 250 words (+/-)

Ice cream products have the highest orders during the evening hours.  Only 17% of orders occur in the 6PM to midnight time period so there is a substantial [lift](https://github.com/brook-miller/mbai-417-data/blob/main/data-models-databases/in-class/2-instacart.ipynb) to ice cream orders during this time frame.  

While this proportion is interesting, the "at least" 5,000 is a bit of an arbitrary cut off.  The results of the 5th query (most commonly ordered products in evening hours) is a valid if less novel angle.

In reality, we'd set up tests and continue to refine.  Given our goal is to create net new orders vs. basket size any marketing efforts should be top of funnel (bringing new or returning customers on to the site) via email, etc... vs. merchandising on the site where featuring ice cream would help to drive basket size.